### PR TITLE
Check ssh-agent liveness before sourcing GPG_ENV

### PR DESCRIPTION
**Pull request by eakret**
Resolves issue #2121

### CODE REVIEW

**Comment by mcornella** [comment]
Hi @eakret, there are pull requests #2063 and #2497 that both change gpg-agent plugin. 
Could you guys (@cdimartino, @blueben) come together and see if you're trying to achieve the same thing? If you *are* trying to solve the same thing, divise a common strategy to solve it. If not, explain to me in layman terms what the differences are, since I don't have a good understanding of what gpg-agent and ssh-agent do or how they are intertwined...

Also, a quick start guide on gpg-agent would be great to test any of these pull requests. Thanks!

**Comment by cdimartino** [comment]
My understanding is that there are two bugs in the script which are causing
similar issues.  My fix https://github.com/robbyrussell/oh-my-zsh/pull/2063
resolves an issue where multiple gpg-agent instances were being spawned due
to a faulty check to see if gpg-agent is already running.  It actually
appears though that the gpg-agent-connect command is now functioning
properly, returning a non-zero exit status if gpg-agent is not running.  My
pull req can be closed if desired.  Either method appears to be a valid way
of checking for a running process now.

**Comment by blueben** [comment]
Long story short, gpg-agent can act as an SSH key agent (--enable-ssh-support) as well as a GPG key agent. That's fine if folks want to use gpg agent for that. However, gpg agent shouldn't try to be an ssh agent if an ssh agent is already running, which is the problem logged in https://github.com/robbyrussell/oh-my-zsh/issues/2496.

My fix has a somewhat more reliable test to determine if an ssh-agent is already running. Doesn't seem to collide with @cdimartino's issue or fix.

**Comment by mcornella** [comment]
Ok, I've had the chance to look into it and I've learned some more. @blueben your PR is simpler and easier to merge, but there's the problem that @eakret mentions in #2121. 

Basically, if ssh-agent is running and our `GPG_ENV` file contains `SSH_*` variables other than current ones (set up by ssh-agent), the check won't return a reliable result because the `SSH_AGENT_PID` will be different than the currently running ssh-agent.

That said, I'm not satisfied with this PR either: it's too complex to understand. A simpler version would be checking for an ssh-agent running before (and after) sourcing the `GPG_ENV` file.

**Comment by eakret** [comment]
Whatever patch is added does need to save away the environmental variables
for the ssh agent before sourcing the GPG_ENV file. If it checks for ssh
agent liveness before sourcing GPG_ENV and ssh agent is alive, but no
longer showing as alive after sourcing GPG_ENV, then it needs to set those
variables back to their previous values.

However, I think the cleanest answer is to remove the ssh liveness check
entirely and just have two plugins: one that starts gpg agent with the
built-in ssh agent functionality and one that just starts the standard gpg
agent.
